compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUselessTypeOperationCallChecker.kt

```diff
@@ -0,0 +1,88 @@
+import org.jetbrains.kotlin.fir.typeContext
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.types.AbstractTypeChecker
+
+// See .../types/CastDiagnosticsUtil.kt for counterparts, including isRefinementUseless, isExactTypeCast, isUpcast.
+object FirUselessTypeOperationCallChecker : FirTypeOperatorCallChecker() {
+    override fun check(expression: FirTypeOperatorCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (expression.operation !in FirOperation.TYPES) return
+        val arg = expression.argument
+
+        val candidateType = arg.typeRef.coneType.upperBoundIfFlexible().fullyExpandedType(context.session)
+        if (candidateType is ConeKotlinErrorType) return
+
+        val targetType = expression.conversionTypeRef.coneType.fullyExpandedType(context.session)
+        if (targetType is ConeKotlinErrorType) return
+
+        // x as? Type <=> x as Type?
+        val refinedTargetType =
+            if (expression.operation == FirOperation.SAFE_AS) targetType.withNullability(ConeNullability.NULLABLE) else targetType
+        val shouldCheckForExactType = when (expression.operation) {
+            FirOperation.IS, FirOperation.NOT_IS -> false
+            // FirOperation.AS, FirOperation.SAFE_AS
+            else -> shouldCheckForExactType(expression, context)
+        }
+        if (isRefinementUseless(context, candidateType, refinedTargetType, shouldCheckForExactType)) {
+            when (expression.operation) {
+                FirOperation.IS, FirOperation.NOT_IS -> {
+                    val negated = expression.operation == FirOperation.NOT_IS
+                    reporter.reportOn(expression.source, FirErrors.USELESS_IS_CHECK, !negated, context)
+                }
+                // FirOperation.AS, FirOperation.SAFE_AS
+                else -> {
+                    reporter.reportOn(expression.source, FirErrors.USELESS_CAST, context)
+                }
+            }
+        }
+    }
+
+    @Suppress("UNUSED_PARAMETER")
+    private fun shouldCheckForExactType(expression: FirExpression, context: CheckerContext): Boolean {
+        // TODO: differentiate if this expression defines the enclosing thing's type
+        //   e.g.,
+        //   val c1 get() = 1 as Number
+        //   val c2: Number get() = 1 <!USELESS_CAST!>as Number<!>
+        return true
+    }
+
+    private fun isRefinementUseless(
+        context: CheckerContext,
+        candidateType: ConeKotlinType,
+        targetType: ConeKotlinType,
+        shouldCheckForExactType: Boolean,
+    ): Boolean {
+        return if (shouldCheckForExactType)
+            isExactTypeCast(candidateType, targetType)
+        else
+            isUpcast(context, candidateType, targetType)
+    }
+
+    private fun isExactTypeCast(candidateType: ConeKotlinType, targetType: ConeKotlinType): Boolean {
+        return candidateType == targetType && candidateType.isExtensionFunctionType == targetType.isExtensionFunctionType
```

Comment:
What does being an extension function have to do with it? Which test data file is testing this? (I don't see any casts to function type anywhere in the diffs.)

Reply:
Hmm... did a search everywhere for cast to extension function type. Is it this? `compiler/testData/diagnostics/tests/cast/ExtensionAsNonExtension.kt`

Reply:
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-extension-function-type/

I added code comments to explain what those are and why we need to check for functional types.